src/core/public/application/types.ts

```diff
@@ -0,0 +1,293 @@
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { Observable, Subject } from 'rxjs';
+
+import { Capabilities } from './capabilities';
+import { ChromeStart } from '../chrome';
+import { IContextProvider } from '../context';
+import { DocLinksStart } from '../doc_links';
+import { HttpStart } from '../http';
+import { I18nStart } from '../i18n';
+import { NotificationsStart } from '../notifications';
+import { OverlayStart } from '../overlays';
+import { PluginOpaqueId } from '../plugins';
+import { UiSettingsClientContract } from '../ui_settings';
+import { RecursiveReadonly } from '../../utils';
+
+/** @public */
+export interface AppBase {
+  id: string;
+
+  /**
+   * The title of the application.
+   */
+  title: string;
+
+  /**
+   * An ordinal used to sort nav links relative to one another for display.
+   */
+  order?: number;
+
+  /**
+   * An observable for a tooltip shown when hovering over app link.
+   */
+  tooltip$?: Observable<string>;
+
+  /**
+   * A EUI iconType that will be used for the app's icon. This icon
+   * takes precendence over the `icon` property.
+   */
+  euiIconType?: string;
+
+  /**
+   * A URL to an image file used as an icon. Used as a fallback
+   * if `euiIconType` is not provided.
+   */
+  icon?: string;
+
+  /**
+   * Custom capabilities defined by the app.
+   */
+  capabilities?: Partial<Capabilities>;
+}
+
+/**
+ * Extension of {@link AppBase | common app properties} with the mount function.
+ * @public
+ */
+export interface App extends AppBase {
+  /**
+   * A mount function called when the user navigates to this app's `rootRoute`.
+   * @param context The mount context for this app.
+   * @param targetDomElement An HTMLElement to mount the application onto.
+   * @returns An unmounting function that will be called to unmount the application.
+   */
+  mount: (context: AppMountContext, params: AppMountParameters) => AppUnmount | Promise<AppUnmount>;
+}
+
+/** @internal */
+export interface LegacyApp extends AppBase {
+  appUrl: string;
+  subUrlBase?: string;
+  linkToLastSubUrl?: boolean;
+}
+
+/**
+ * The context object received when applications are mounted to the DOM.
+ * @public
+ */
+export interface AppMountContext {
+  /**
+   * Core service APIs available to mounted applications.
+   */
+  core: {
```

Comment:
Is this different from `CoreStart`? Maybe it could be

```ts
core: CoreStart;
```

Reply:
It is the same currently, but that will change in the very near future. For instance, part of this context will include an API for blocking navigation to other apps that will not be available in CoreStart but will be available in `context.core.application.blockNavigation`